js/ui/payg.js

```diff
@@ -38,7 +38,11 @@ const SPINNER_ANIMATION_TIME_MSECS = 300;
 var SPINNER_ICON_SIZE_PIXELS = 16;
 
 const NOTIFICATION_TITLE_TEXT = _('Pay As You Go');
-const NOTIFICATION_EARLY_CODE_ENTRY_TEXT = _('Enter a 14-digit keycode to extend the time before your credit expires.');
+
+// These will be passed to ngettext to get pluralized translations
+const NOTIFICATION_EARLY_CODE_ENTRY_TEXT_1 = 'Enter a new keycode (%s character) to extend the time before your credit expires.';
+const NOTIFICATION_EARLY_CODE_ENTRY_TEXT_N = 'Enter a new keycode (%s characters) to extend the time before your credit expires.';
```

Comment:
I’m not 100% sure, but I suspect these strings should be marked with `gettext_noop()` (in C this is `N_()`) or something like that, so that `xgettext` knows to extract them from the source code into the `.pot` file at build time.

There are two parts to using gettext: build-time extraction of the translatable strings into the `.pot` file (which translators then localise as `$lang.po` files), and runtime lookup of the translated forms of those strings from the translation catalogue. Extraction of the strings uses some basic source code parsing, looking for strings being passed to certain gettext functions, and assuming all of them are to be translated. I don’t think it’s clever enough to fold constants.

Normally the `_()` macro (equivalently the `gettext()` function) handles both parts, but in the case of defining the translatable strings separately from where they’re translated using `ngettext()`, the strings need to be explicitly marked as for extraction into a `.pot` file.

Reply:
That makes sense, thanks for the explanation (I never dealt much with localization before). I skimmed through https://github.com/guillaumepotier/gettext.js/ and did not find anything equivalent to `N_()` or `gettext_noop`. I think the safe approach is to just move the definition of these strings to where they are used, together with NOTIFICATION_TITLE_TEXT and NOTIFICATION_DETAILED_FORMAT_STRING -- while it is nice to have them grouped in the top of the file, we're not gaining much with these macros, as they are only used once.

Reply:
> I think the safe approach is to just move the definition of these strings to where they are used, together with NOTIFICATION_TITLE_TEXT and NOTIFICATION_DETAILED_FORMAT_STRING -- while it is nice to have them grouped in the top of the file, we're not gaining much with these macros, as they are only used once.

Yeah, that seems pragmatic

Reply:
I don't think that's the gettext module exposed by GJS. I think it's this one, which is just a thin wrapper around the corresponding functions in glib. https://gitlab.gnome.org/GNOME/gjs/-/blob/master/modules/core/_gettext.js